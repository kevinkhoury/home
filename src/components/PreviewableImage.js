import React, { Fragment } from 'react'
import Img from 'gatsby-image'
import PropTypes from 'prop-types'

const PreviewableImage = ({ src, isPreview, alt, caption }) => {
  const sources =
    !!isPreview || typeof src === 'string'
      ? null
      : !!src.childImageSharp
      ? src.childImageSharp.fluid
      : !!src.m && !!src.d
      ? [
          src.m.childImageSharp.fluid,
          {
            ...src.d.childImageSharp.fluid,
            media: `(min-width: 768px)`,
          },
        ]
      : null
  return (
    <Fragment>
      {!!sources ? (
        <Img className="gatsby-resp-image-image" fluid={sources} alt={alt} />
      ) : (
        <img className="gatsby-resp-image-image" src={src} alt={alt} />
      )}
      {!!caption && <figcaption>{caption}</figcaption>}
    </Fragment>
  )
}

PreviewableImage.propTypes = {
  src: PropTypes.oneOfType([PropTypes.string, PropTypes.object]).isRequired,
  isPreview: PropTypes.bool,
  alt: PropTypes.string.isRequired,
  caption: PropTypes.string,
}

export default PreviewableImage
