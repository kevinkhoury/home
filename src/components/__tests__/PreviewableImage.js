import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import PreviewableImage from '../PreviewableImage'

const M_D_IMG = {
  d: {
    childImageSharp: {
      fluid: {
        foobar: `desktop fluid`,
      },
    },
  },
  m: {
    childImageSharp: { fluid: { foobar: `mobile fluid` } },
  },
}

const DEFAULT_PROPS = {
  src: {
    childImageSharp: {
      fluid: { foobar: `basic fluid` },
      original: {
        height: 100,
        width: 200,
      },
    },
  },
  isPreview: false,
  alt: 'alttext',
  caption: 'captiontext',
}

jest.mock('gatsby-image', () => (props) => (
  <div data-testid="gatsby-image">{JSON.stringify(props)}</div>
))

afterEach(() => {
  jest.clearAllMocks()
})

describe('PreviewableImage', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<PreviewableImage {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('creates a gatsby image if given childImageSharp', () => {
    render(<PreviewableImage {...DEFAULT_PROPS} />)
    const img = screen.getByTestId('gatsby-image')
    expect(img.innerHTML).toMatch(/"className":"gatsby-resp-image-image"/)
  })
  it('creates a caption if given', () => {
    render(<PreviewableImage {...DEFAULT_PROPS} />)
    const caption = screen.getByText(/captiontext/)
    expect(caption).toBeTruthy()
  })
  it('combines desktop and mobile image sources if given', () => {
    const props = { ...DEFAULT_PROPS, src: M_D_IMG }
    render(<PreviewableImage {...props} />)
    const img = screen.getByTestId('gatsby-image')
    expect(img.innerHTML).toMatch(/desktop fluid/)
    expect(img.innerHTML).toMatch(/mobile fluid/)
  })
  it('creates a normal image if given a string', () => {
    render(<PreviewableImage src="/foobar.jpg" alt="alttext" />)
    const img = screen.getByAltText('alttext')
    expect(img.getAttribute('src')).toBe('/foobar.jpg')
  })
})
