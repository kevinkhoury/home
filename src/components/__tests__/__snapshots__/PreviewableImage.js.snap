// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewableImage snapshot renders properly 1`] = `
Array [
  <div
    data-testid="gatsby-image"
  >
    {"className":"gatsby-resp-image-image","fluid":{"foobar":"basic fluid"},"alt":"alttext"}
  </div>,
  <figcaption>
    captiontext
  </figcaption>,
]
`;
